### ISSUE TITLE
def_node_matcher for a chain of calls?

### ISSUE THREAD

**Issue by andyw8** [rubocop-hq/rubocop#6715]
I have this node matcher which will match code such as `foo.stubs(:bar).with(:baz)`:

```
def_node_matcher :candidate?, <<-CODE
  $(send
    $(send
      _
      ${:stubs :expects}
      _
    )
    :with
    ...
  )
CODE
```

I now want to extend this to match `foo.stubs(...).with(...).returns(...)`.

I've been reading through the Node Pattern docs and trying different things to try achieve this. I started with a naive guess:

```
def_node_matcher :candidate?, <<-CODE
  $(send
    $(send
      _
      ${:stubs :expects}
      _
    )
    :with
    ...
    :returns
    ...
  )
CODE
```

but I think that's probably not the right way.

I've been looking through the docs and trying a few variations but without luck. I've been using `ruby-parse -e` to see the the s-expresssions, but I'm new to the node matcher pattern and I feel I don't have a good grasp of it yet.

Any tips on how I can make this work? Or a cop I can look at for inspiration?

**Comment by Drenmi** [comment]
Not entirely sure I understand the requirement correctly, but you could try starting with this:

```
(send (send (send _ :stubs _) :with _) :returns _)
```

This will match a method chain of `#stubs`, `#with`, and `#returns` with any arguments, sent to any receiver. The wild cards appear in the same order in the pattern as they would in your example:

```ruby
_.stubs(_).with(_).returns(_)
```

Does this help? 🙂

**Comment by andyw8** [comment]
@Drenmi Thank you, that was very helpful, I'm making progress now.

**Comment by Drenmi** [comment]
Great! I'll close the issue for now. Ping me again if you need any guidance. 🙇
